pypdf/_reader.py

```diff
@@ -286,29 +290,30 @@ def _get_page_number_by_indirect(
         self, indirect_reference: Union[None, int, NullObject, IndirectObject]
     ) -> Optional[int]:
         """
-        Generate _page_id2num.
+        Retrieve the page number from an indirect reference.
 
         Args:
-            indirect_reference:
+            indirect_reference: The indirect reference to locate.
 
         Returns:
-            The page number or None
+            Page number or None.
         """
         if self._page_id2num is None:
             self._page_id2num = {
-                x.indirect_reference.idnum: i for i, x in enumerate(self.pages)  # type: ignore
+                x.indirect_reference.idnum: i
+                for i, x in enumerate(self.pages)
+                if x.indirect_reference is not None
```

Comment:
you should use is_null_none()
```suggestion
                if is_null_or_none(x.indirect_reference)
```